### DISCUSSION TITLE
How do I use terratest to write automated tests on resources that are being added to a custom provider without making a release ?

### DISCUSSION THREAD

**Discussion by nirav-rafay** [gruntwork-io/knowledge-base#727] [How do I]
Some more details

1) let's say there is a custom terraform provider - x
2) I have added new resources to provider 'x', it is under development.
3) At this point I would like to use terratest and write an automated test suite to ensure everything is working as expected against this **unreleased provider x**

---

<ins datetime="2023-06-06T07:40:03Z">
  <p><a href="https://support.gruntwork.io/hc/requests/110230">Tracked in ticket #110230</a></p>
</ins>

**Comment by denis256** [comment]
Hi,
from what is mentioned, looks like can be used `terraform` module to run `plan/apply` and checked results of execution:
* in case of `plan` can be readded and verified plan output
* in case of `apply` use Go client to verify if expected actions were performed 

Examples:
https://github.com/gruntwork-io/terratest/blob/master/test/terraform_aws_example_plan_test.go#L17
https://github.com/gruntwork-io/terratest/blob/master/test/terraform_aws_example_test.go

**Comment by nirav-rafay** [comment]
hey @denis256, thanks for your response. In above example I am assuming the provider aws has a resource "aws_instance" that is already available and released ( present in terraform registry ). What I am looking for is something like to replicate below where my newly added resources / modules are yet to be released and ability to overrride that using my local directory

```
provider_installation {

  # Use /home/developer/tmp/terraform-null as an overridden package directory
  # for the hashicorp/null provider. This disables the version and checksum
  # verifications for this provider and forces Terraform to look for the
  # null provider plugin in the given directory.
  dev_overrides {
    "hashicorp/null" = "/home/developer/tmp/terraform-null"
  }

  # For all other providers, install them directly from their origin provider
  # registries as normal. If you omit this, Terraform will _only_ use
  # the dev_overrides block, and so no other providers will be available.
  direct {}
}
``` 

ref: https://developer.hashicorp.com/terraform/cli/config/config-file#development-overrides-for-provider-developers

**Reply by denis256** [reply]
At this point, I think can be used 2 approaches:
 * pass `TF_CLI_CONFIG_FILE` environment variable in `terraform.Options` pointing on file with `provider_installation` configuration https://github.com/gruntwork-io/terratest/blob/master/modules/terraform/options.go#L57
 * use `files.CopyTerraformFolderToTemp` to copy tested directory to a temporary location, `provider_installation` file in the temporary directory
